storage/cache/subtree_cache.go

```diff
@@ -149,30 +149,34 @@ func (s *SubtreeCache) stratumInfoForPrefixLength(l int) stratumInfo {
 // passed in node IDs, uses getSubtrees to retrieve them, and finally populates
 // the cache structures with the data.
 func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc) error {
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-
-	// Figure out the set of subtrees we need:
-	want := make(map[string]*storage.NodeID)
-	for _, id := range ids {
-		id := id
-		sInfo := s.stratumInfoForNodeID(id)
-		pxKey := id.PrefixAsKey(sInfo.prefixBytes)
-		// TODO(al): fix for non-uniform strata
-		id.PrefixLenBits = sInfo.prefixBytes * depthQuantum
-		if _, ok := s.subtrees[pxKey]; !ok {
-			want[pxKey] = &id
+	// Figure out the set of subtrees we need.
+	want := make(map[string]storage.NodeID)
+	func() {
+		s.mutex.RLock()
+		defer s.mutex.RUnlock()
+
+		for _, id := range ids {
+			sInfo := s.stratumInfoForNodeID(id)
+			pxKey := id.PrefixAsKey(sInfo.prefixBytes)
+			// TODO(al): Fix for non-uniform strata.
+			id.PrefixLenBits = sInfo.prefixBytes * depthQuantum
+			if _, ok := s.subtrees[pxKey]; !ok {
```

Comment:
I assume the lines above this that are effectively statically mapping a node ID to a strata key are super fast. If that isn't the case, the creation of the set of all required strata could happen outside the critical block, and then the subtraction of the subtrees already in the cache could be the only part inside the mutex.

Reply:
It's not *super* fast, as it involves one string copy, but it doesn't look horrible.